tools/render-test/shader-input-layout.cpp

```diff
@@ -1100,12 +1126,82 @@ namespace renderer_test
         return SLANG_OK;
     }
 
+    template<typename T>
+    void generateTextureDataWithTargetTStorage(TextureData& output, const InputTextureDesc& desc, gfx::FormatInfo& formatInfo)
+    {
+        TextureData work;
+        generateTextureDataRGB8(work, desc);
+
+        output.init(desc.format);
+
+        output.m_textureSize = work.m_textureSize;
+        output.m_mipLevels = work.m_mipLevels;
+        output.m_arraySize = work.m_arraySize;
+
+        List<TextureData::Slice>& dstSlices = output.m_slices;
+
+        Index numSlices = work.m_slices.getCount();
+        dstSlices.setCount(numSlices);
+
+        for (int i = 0; i < numSlices; ++i)
+        {
+            const TextureData::Slice& srcSlice = work.m_slices[i];
+
+            const Index pixelCount = srcSlice.valuesCount;
+            const uint8_t* srcPixels = (const uint8_t*)srcSlice.values;
 
+            T* dstPixels = (T*)output.setSliceCount(i, pixelCount);
+            switch (formatInfo.channelCount)
+            {
+            case 1:
+            {
+                for (Index j = 0; j < pixelCount; ++j, srcPixels += 4, dstPixels++)
+                {
+                    // Copy out r
+                    dstPixels[0] = T(srcPixels[0] * (1.0f / 255));
```

Comment:
I am not sure if this is a right way for integer types.
The original code looks like the assumption is that srcPixels is 8 bit format whose value was normalized.
And it is converting to a normalized floating value.
As an example, when srcPixels[0] was 254, dstPixels[0] would get 0.99f as a floating value, which makes sense.
But if the destination type is an integertype and srcPixels[0] is 254, dstPixels[0] will get 0 because int(254 / 255) is zero. The resulting value will be always zero or one, which doesn't make sense.

I think that when the destination is an integer value, you need to multiply the result with the max value of the type to keep it normalized.
Something like this,
```
dstPixels[0] = T(srcPixels[0] * limits<T>::max() / 255);
```

Reply:
the code assumes the type is 0 or 1, anything in-between does not matter

the code during initialization only has: one, zero, one & zero, as the initialization parameters, as a result the code works for the purposes of testing

Reply:
If the values in srcPixels are expected to be zero or one, why do we need to divide them with 255?
And is the input values never going to be changed to values other than zero and one?

The intention here looks to be converting the image format from RGBA8 to R64_UINT, as an example.
I think you need to do a proper conversion.

Reply:
0 or 1 was a unorm description, I explained the process with unorm because the data is actually stored as uint8_t, 1/255*255 gives a float representation (for casting) which is how floatToHalf works. The existing system was only designed to handle 0 or 1, any other value and you are at the mercy of IEEE and floating point inaccuracies

a proper conversion would mean

Reply:
thinking about a proper conversion:

Reply:
lets say our input system says: load zeroPointFive, this would require the code before this which is loading the initial data to setup 128,

lets say we input two, this would require the code before this which is loading the initial data to setup 2 since 512 is not an option

Reply:
I don't understand what you are saying.
Here is how I see it.
When we convert from RGBA8 format to RGBA32F, we do the following conversion.
```
rgba32f[0] = float(rgba8[0]) / 255.f;
```
When we convert from RGBA8 to RGBA16, as an example, we do the following,
```
rgba16[0] = rgba8[0] << 8;
```
Your implementation works for only one case where the input is 8bit integer and output is a floating point. It wouldn't work when the output is an integer format.

I don't know what you mean by 1.
It could mean a near dark color in RGBA8 format. (1/255).
Or it could mean a full white color in RGBA8 format (255/255).
Your expression is very hard to understand even with the examples.

Reply:
my implementation worked for all types as long as it was 0 or 1 as input. It simply took a float which was expected to be 0.0f or 1.0f, and casted it to another type.

I changed the implementation (as per the request) to work with non 0/1 inputs as well for the future.

about the new implementation: I did not want to make code that did something in the nature of: `rgba16[0] = rgba8[0] << 8;` since then when adding all 12 types we would have a big bundle of code which would be hard to iterate on, change, and reuse. I wanted to keep everything easy to understand and modify to allow easy additions of functionality. I think the new one is quite easy to work around.